# **Ensemble Learning (Bagging) Exercise**

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import pandas as pd

df = pd.read_csv('heart.csv')
df

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0      172              N      0.0       Up             0  
1      156              N      1.0     Flat             1  
2       98              N      0.0       Up             0  
3      108              Y      1.5     Flat             1  
4      122              N      0.0       Up             0  
..     ...            ...      ...      ...           ...  
913    132              N      1.2     Flat             1  
914    141              N      3.4     Flat             1  
915    115              Y      1.2     Flat             1  
916    174              N      0.0     Flat             1  
917    173              N      0.0       Up             0  

[918 rows x 12 columns]

In [3]:
df.isna().sum()

Age               0
Sex               0
ChestPainType     0
RestingBP         0
Cholesterol       0
FastingBS         0
RestingECG        0
MaxHR             0
ExerciseAngina    0
Oldpeak           0
ST_Slope          0
HeartDisease      0
dtype: int64

In [4]:
df.describe()

Age   RestingBP  Cholesterol   FastingBS       MaxHR  \
count  918.000000  918.000000   918.000000  918.000000  918.000000   
mean    53.510893  132.396514   198.799564    0.233115  136.809368   
std      9.432617   18.514154   109.384145    0.423046   25.460334   
min     28.000000    0.000000     0.000000    0.000000   60.000000   
25%     47.000000  120.000000   173.250000    0.000000  120.000000   
50%     54.000000  130.000000   223.000000    0.000000  138.000000   
75%     60.000000  140.000000   267.000000    0.000000  156.000000   
max     77.000000  200.000000   603.000000    1.000000  202.000000   

          Oldpeak  HeartDisease  
count  918.000000    918.000000  
mean     0.887364      0.553377  
std      1.066570      0.497414  
min     -2.600000      0.000000  
25%      0.000000      0.000000  
50%      0.600000      1.000000  
75%      1.500000      1.000000  
max      6.200000      1.000000

### **Treating Outliers:**


**Outliers can distort model performance (e.g., increase error in machine learning).**

The formula mean ± 3 * standard deviation (μ ± 3σ) is based on the 68-95-99.7 rule (also called the Empirical Rule) in statistics.

This rule states that for a normally distributed dataset:

    68% of the data falls within 1 standard deviation (μ ± 1σ).

    95% of the data falls within 2 standard deviations (μ ± 2σ).

    99.7% of the data falls within 3 standard deviations (μ ± 3σ).
By removing values beyond 3 standard deviations, we are removing only 0.3% of extreme values (outliers) while keeping 99.7% of the data intact

In [5]:
df[df.RestingBP > (df.RestingBP.mean() + 3 * df.RestingBP.std())]

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
109   39   M           ATA        190          241          0     Normal   
241   54   M           ASY        200          198          0     Normal   
365   64   F           ASY        200            0          0     Normal   
399   61   M           NAP        200            0          1         ST   
592   61   M           ASY        190          287          1        LVH   
732   56   F           ASY        200          288          1        LVH   
759   54   M           ATA        192          283          0        LVH   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
109    106              N      0.0       Up             0  
241    142              Y      2.0     Flat             1  
365    140              Y      1.0     Flat             1  
399     70              N      0.0     Flat             1  
592    150              Y      2.0     Down             1  
732    133              Y      4.0     Down             1  
759    195              N      0.0       Up             1

In [6]:
df_1 = df[df.RestingBP <= (df.RestingBP.mean() + 3 * df.RestingBP.std())]
df_1.shape

(911, 12)

In [7]:
df[df.Cholesterol > (df.Cholesterol.mean() + 3 * df.Cholesterol.std())]

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
76    32   M           ASY        118          529          0     Normal   
149   54   M           ASY        130          603          1     Normal   
616   67   F           NAP        115          564          0        LVH   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
76     130              N      0.0     Flat             1  
149    125              Y      1.0     Flat             1  
616    160              N      1.6     Flat             0

In [8]:
df_2 = df_1[df.Cholesterol <= (df.Cholesterol.mean() + 3 * df.Cholesterol.std())]
df_2.shape

(908, 12)

In [9]:
df[df.MaxHR > (df.MaxHR.mean() + 3 * df.MaxHR.std())]

Empty DataFrame
Columns: [Age, Sex, ChestPainType, RestingBP, Cholesterol, FastingBS, RestingECG, MaxHR, ExerciseAngina, Oldpeak, ST_Slope, HeartDisease]
Index: []

In [10]:
df[df.Oldpeak > (df.Oldpeak.mean() + 3 * df.Oldpeak.std())]

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
166   50   M           ASY        140          231          0         ST   
702   59   M            TA        178          270          0        LVH   
771   55   M           ASY        140          217          0     Normal   
791   51   M           ASY        140          298          0     Normal   
850   62   F           ASY        160          164          0        LVH   
900   58   M           ASY        114          318          0         ST   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
166    140              Y      5.0     Flat             1  
702    145              N      4.2     Down             0  
771    111              Y      5.6     Down             1  
791    122              Y      4.2     Flat             1  
850    145              N      6.2     Down             1  
900    140              N      4.4     Down             1

In [11]:
df_3 = df_2[df.Oldpeak <= (df.Oldpeak.mean() + 3 * df.Oldpeak.std())]
df_3.shape

(902, 12)



> **We dropped 16 otliers.**



### **Handling Dummy Variables:**

In [12]:
df.Sex.unique()

array(['M', 'F'], dtype=object)

In [13]:
df.ChestPainType.unique()

array(['ATA', 'NAP', 'ASY', 'TA'], dtype=object)

In [14]:
df.FastingBS.unique()

array([0, 1])

In [15]:
df.RestingECG.unique()

array(['Normal', 'ST', 'LVH'], dtype=object)

In [16]:
df.ExerciseAngina.unique()

array(['N', 'Y'], dtype=object)

In [17]:
df.ST_Slope.unique()

array(['Up', 'Flat', 'Down'], dtype=object)

In [18]:
df_4 = df_3.copy()

df_4.ExerciseAngina.replace({'N':0, 'Y':1}, inplace=True)

df_4.RestingECG.replace({'Normal':1, 'ST':2, 'LVH':3}, inplace=True)

df_4.ST_Slope.replace({'Down':1, 'Flat':2, 'Up':3}, inplace=True)

df_4.Sex.replace({'M':1, 'F':0}, inplace=True)

df_4.head()

Age  Sex ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0   40    1           ATA        140          289          0           1   
1   49    0           NAP        160          180          0           1   
2   37    1           ATA        130          283          0           2   
3   48    0           ASY        138          214          0           1   
4   54    1           NAP        150          195          0           1   

   MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
0    172               0      0.0         3             0  
1    156               0      1.0         2             1  
2     98               0      0.0         3             0  
3    108               1      1.5         2             1  
4    122               0      0.0         3             0

In [19]:
df_5 = pd.get_dummies(df_4, drop_first=True)
df_5.head()

Age  Sex  RestingBP  Cholesterol  FastingBS  RestingECG  MaxHR  \
0   40    1        140          289          0           1    172   
1   49    0        160          180          0           1    156   
2   37    1        130          283          0           2     98   
3   48    0        138          214          0           1    108   
4   54    1        150          195          0           1    122   

   ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  ChestPainType_ATA  \
0               0      0.0         3             0               True   
1               0      1.0         2             1              False   
2               0      0.0         3             0               True   
3               1      1.5         2             1              False   
4               0      0.0         3             0              False   

   ChestPainType_NAP  ChestPainType_TA  
0              False             False  
1               True             False  
2              False             False  
3              False             False  
4               True             False

### **Splitting features (X) and labels (y):**

In [20]:
X = df_5.drop('HeartDisease', axis='columns')
y = df_5.HeartDisease

### **Scaling the features (This helps the model train faster and more accurately):**

In [21]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
X_scaled

array([[-1.42896269,  0.51485643,  0.46089071, ...,  2.06757196,
        -0.53547478, -0.22914788],
       [-0.47545956, -1.94228905,  1.5925728 , ..., -0.4836591 ,
         1.86750159, -0.22914788],
       [-1.74679706,  0.51485643, -0.10495034, ...,  2.06757196,
        -0.53547478, -0.22914788],
       ...,
       [ 0.37209878,  0.51485643, -0.10495034, ..., -0.4836591 ,
        -0.53547478, -0.22914788],
       [ 0.37209878, -1.94228905, -0.10495034, ...,  2.06757196,
        -0.53547478, -0.22914788],
       [-1.64085227,  0.51485643,  0.3477225 , ..., -0.4836591 ,
         1.86750159, -0.22914788]])

### **Training a Classification Model Using Support Vector Machine (SVM):**

In [22]:
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score

svm_scores = cross_val_score(SVC(), X_scaled, y, cv=5)
svm_scores.mean()

np.float64(0.8159054634745242)

### **Using Bagging with SVM:**

In [23]:
from sklearn.ensemble import BaggingClassifier

bag_model = BaggingClassifier(
    estimator=SVC(),
    n_estimators=100, # number of subset of dataset (100 model)
    max_samples=0.8,  # number of samples in each subset. Here we use 80% of dataset for each sample
    oob_score=True,   # use samples those are out of subsets (bags) for test the accuracy of models
    random_state=0    # to make sure that we always get the same result every time we run the code.
    )
bag_svm_scores = cross_val_score(bag_model, X_scaled, y, cv=5)
bag_svm_scores.mean()

np.float64(0.8159116022099446)

### **Training a Classification Model Using Decision Tree Classifier:**

In [24]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score

dt_scores = cross_val_score(DecisionTreeClassifier(), X_scaled, y, cv=5)
dt_scores.mean()

np.float64(0.7238489871086556)

### **Using Bagging with Decision Tree Classifier:**

In [25]:
from sklearn.ensemble import BaggingClassifier

bag_model = BaggingClassifier(
    estimator=DecisionTreeClassifier(),
    n_estimators=100, # number of subset of dataset (100 model)
    max_samples=0.8,  # number of samples in each subset. Here we use 80% of dataset for each sample
    oob_score=True,   # use samples those are out of subsets (bags) for test the accuracy of models
    random_state=0    # to make sure that we always get the same result every time we run the code.
    )

bagging_scores = cross_val_score(bag_model, X_scaled, y, cv=5)
bagging_scores.mean()

np.float64(0.7992694904849602)

### **Trying a Random Forest model:**

A Random Forest is an advanced version of bagging with decision trees. It not only uses different subsets of data but also chooses random features at each split.

In [26]:
from sklearn.ensemble import RandomForestClassifier

rf_scores = cross_val_score(RandomForestClassifier(), X_scaled, y, cv=5)
rf_scores.mean()

np.float64(0.8236648250460405)



---



***As we can see from the average scores of each model, random forest performs better with about 82% accuracy :)***